buildSrc/src/main/kotlin/com/datadog/gradle/plugin/jsonschema/generator/KotlinSpecGenerator.kt

```diff
@@ -0,0 +1,114 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.gradle.plugin.jsonschema.generator
+
+import com.datadog.gradle.plugin.jsonschema.TypeDefinition
+import com.datadog.gradle.plugin.jsonschema.asKotlinTypeName
+import com.squareup.kotlinpoet.ANY
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.LIST
+import com.squareup.kotlinpoet.NOTHING
+import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
+import com.squareup.kotlinpoet.SET
+import com.squareup.kotlinpoet.TypeName
+
+abstract class KotlinSpecGenerator<I : Any, O : Any>(
+    private val packageName: String,
+    val nestedTypes: MutableSet<TypeDefinition>,
+    val knownTypeNames: MutableSet<String>
+) {
+
+    /**
+     * Generates a KotlinPoet Spec based on the provided Json information
+     * @param rootTypeName the name of the root type (matching the name of the file)
+     * @param definition the definition read from a Json Schema file
+     */
+    abstract fun generate(definition: I, rootTypeName: String): O
+
+    // region Utilities
+
+    protected fun TypeDefinition.name(): String? {
+        return when (this) {
+            is TypeDefinition.Array,
+            is TypeDefinition.Primitive,
+            is TypeDefinition.Constant,
+            is TypeDefinition.Null -> null
+            is TypeDefinition.Enum -> name
+            is TypeDefinition.Class -> name
+            is TypeDefinition.MultiClass -> name
+        }
+    }
+
+    protected fun TypeDefinition.asKotlinTypeName(
+        rootTypeName: String
+    ): TypeName {
+        return when (this) {
+            is TypeDefinition.Null -> NOTHING
+            is TypeDefinition.Primitive -> type.asKotlinTypeName()
+            is TypeDefinition.Constant -> type.asKotlinTypeName()
+            is TypeDefinition.Class -> {
+                val def = withUniqueTypeName()
+                nestedTypes.add(def)
+                ClassName(packageName, rootTypeName, def.name)
+            }
+            is TypeDefinition.Array -> {
+                if (uniqueItems) {
+                    SET.parameterizedBy(items.asKotlinTypeName(rootTypeName))
+                } else {
+                    LIST.parameterizedBy(items.asKotlinTypeName(rootTypeName))
+                }
+            }
+            is TypeDefinition.Enum -> {
+                val def = withUniqueTypeName()
+                nestedTypes.add(def)
+                ClassName(packageName, rootTypeName, def.name)
+            }
+            is TypeDefinition.MultiClass -> {
```

Comment:
we are missing the case where the `ref` is defined outside of the current `schema` in a separated one. In that case it needs to be `ClassName(packageName, name)`.
See my example in the comments section of this PR.